core/src/saros/versioning/VersionManager.java

```diff
@@ -1,330 +1,184 @@
-      }
-
-      try {
-        remoteCompatibility = Compatibility.fromCode(Integer.valueOf(compatibilityString));
-      } catch (NumberFormatException e) {
-        LOG.warn("remote compatibility string contains non numerical characters");
-      }
-
-      compatibility = determineCompatibility(localVersion, remoteVersion);
-
-      // believe what the remote side told us in case we are too old
-      if (compatibility == Compatibility.TOO_OLD && remoteCompatibility == Compatibility.OK)
-        compatibility = Compatibility.OK;
+  public VersionCompatibilityResult determineVersionCompatibility(JID rqJID) {
+    Version remoteVersion =
+        contactsService
+            .getContact(rqJID.getRAW())
+            .map(remoteInfo::get)
+            .map(ClientInfo::getVersion)
+            .orElse(null);
+    if (remoteVersion == null) {
+      log.warn("remote version not found for: " + rqJID);
+      return new VersionCompatibilityResult(Compatibility.UNKNOWN, localVersion, Version.INVALID);
     }
 
+    Compatibility compatibility = determineCompatibility(localVersion, remoteVersion);
     return new VersionCompatibilityResult(compatibility, localVersion, remoteVersion);
   }
 
   /**
-   * Sets an compatibility char that contains additional version information. The chart should be
-   * loaded from a property file which must use the following syntax:
-   *
-   * <pre>
-   *     <tt>local_version = remote_version { & remote_version }</tt>
-   * </pre>
+   * Set a Information that should be broadcasted to other Contacts.
    *
-   * @param chart the chart to set or <code>null</code> to reset the compatibility chart
+   * @param key
+   * @param value if null key is removed from local info
    */
-  public void setCompatibilityChart(Properties chart) {
-
-    final Map<Version, List<Version>> newCompatibilityChart = new HashMap<Version, List<Version>>();
-
-    if (chart == null) chart = new Properties(); // dummy
-
-    for (final Object versionKey : chart.keySet()) {
-      final Version version = Version.parseVersion(versionKey.toString());
-
-      if (version == Version.INVALID) continue;
-
-      final List<Version> compatibleVersions = new ArrayList<Version>();
-
-      for (final String compatibleVersionString : chart.get(versionKey).toString().split("&")) {
-
-        final Version compatibleVersion = Version.parseVersion(compatibleVersionString.trim());
-
-        if (compatibleVersion != Version.INVALID) compatibleVersions.add(compatibleVersion);
-      }
-
-      if (!compatibleVersions.contains(version)) compatibleVersions.add(version);
-
-      newCompatibilityChart.put(version, compatibleVersions);
-    }
-
-    final Version currentVersion = localVersion;
+  public void setLocalInfo(String key, String value) {
+    Objects.requireNonNull(key, "key is null");
+
+    if (value == null) localInfo.remove(key);
+    else localInfo.put(key, value);
+
+    if (remoteInfo.isEmpty()) return;
+    dispatchThread.executeAsDispatch(
```

Comment:
I just looked till
https://github.com/saros-project/saros/blob/f13bdc3ea5c5d295428c01db145bd8d5a41e189d/core/src/saros/net/internal/XMPPTransmitter.java#L131

and thought that synchronized doesn't look good.

I was thinking that I don't want to hang the ui on changing for example the skype name in prefs and remembered you mentioning this https://github.com/saros-project/saros/pull/630#discussion_r312391180

Could you explain where do you see something leaking, I don't see what you mean by internal stuff.

Reply:
```sendPacket``` do not block even it is synchronized. The only reason the DispatchThread is visible outside the network package is so skip serialization if you send messages to yourself (the only access in the ActitvitySequencer). The only thing that had to be synchronized is the setting and sending. Otherwise you could trigger a send on contact update, generate a packet which is not send yet, then you set something and this packet is transmitted and afterwards the old packet is transmitted (guess that will never happen but theoretically it could).

Reply:
Ok, I removed the DispatchThread.